astronomer/houston-api
# Move reconnect logic into Publisher

Fixes https://github.com/astronomer/issues/issues/1609

**To Test**
Steps to test the behavior in Local Dev:

1. `docker-compose up -d` to get nat-streaming running
2. Start Houston-API and Astro-UI
3. In another terminal window `docker container ls` and note the container ID for nats-streaming
4. `docker container stop {CONTAINER_ID}`
5. You should see Houston-API disconnect message in the logs for Houston
6. Quickly `docker-compose up -d` again to bring NATS back up
7. You should see Houston-API reconnected correctly
8. Go to localhost:5000
9. Create a new deployment (or visit and existing deployment)
10. Scroll down and select delete deployment and press OK on the pop up
11. The deployment should correctly be removed

**Expected behavior**
You should be able to repeat the above reproduce steps (3-11) as much as you would like without seeing any issues.  The only potential issue is if the max reconnect attempts is reached (we should alert and fail health checks).

**Terminal Output (with the fix in place)**

During Houston startup: 
```
2020-07-31T13:04:14 INFO Connecting to NATS cluster houston using client houston-delete-deployment-e279f232-c6c7-43ce-b962-93d1fb6757c1 and the url nats://localhost:4222
2020-07-31T13:04:15 INFO Server ready at http://localhost:8871/v1
2020-07-31T13:04:15 INFO Subscriptions ready at ws://localhost:8871/ws
2020-07-31T13:04:15 INFO Connected to NATS cluster houston using client houston-delete-deployment-e279f232-c6c7-43ce-b962-93d1fb6757c1 and url nats://localhost:4222
2020-07-31T13:04:16 DEBUG POST /v1 200 210.627 ms - 88346
```

During Reconnect Houston-API will log similar to:
```
2020-07-31T12:09:34 INFO Attempting to reconnect to cluster houston using client houston-delete-deployment-aebb26a5-0762-42ce-8572-f5f8791fdde6 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Attempting to reconnect to cluster houston using client houston-create-deployment-7da414e1-4466-4bdc-9139-5c37fb837582 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Reconnected to cluster houston using client houston-delete-deployment-aebb26a5-0762-42ce-8572-f5f8791fdde6 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Connecting to NATS cluster houston using client houston-delete-deployment-eb91e506-911c-4709-8d91-dbb4f2c09160 and the url nats://localhost:4222
2020-07-31T12:09:34 INFO Reconnected to cluster houston using client houston-create-deployment-7da414e1-4466-4bdc-9139-5c37fb837582 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Connecting to NATS cluster houston using client houston-create-deployment-0d48d6e9-a957-4b7c-8a10-507d3d7cd0a0 and the url nats://localhost:4222
2020-07-31T12:09:34 INFO Connection lost for cluster houston using client houston-delete-deployment-aebb26a5-0762-42ce-8572-f5f8791fdde6 and url nats://localhost:4222
2020-07-31T12:09:34 ERROR client has been replaced or is no longer registered
2020-07-31T12:09:34 INFO Closed connection for cluster houston using client houston-delete-deployment-aebb26a5-0762-42ce-8572-f5f8791fdde6 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Connection lost for cluster houston using client houston-create-deployment-7da414e1-4466-4bdc-9139-5c37fb837582 and url nats://localhost:4222
2020-07-31T12:09:34 ERROR client has been replaced or is no longer registered
2020-07-31T12:09:34 INFO Closed connection for cluster houston using client houston-create-deployment-7da414e1-4466-4bdc-9139-5c37fb837582 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Connected to NATS cluster houston using client houston-delete-deployment-eb91e506-911c-4709-8d91-dbb4f2c09160 and url nats://localhost:4222
2020-07-31T12:09:34 INFO Connected to NATS cluster houston using client houston-create-deployment-0d48d6e9-a957-4b7c-8a10-507d3d7cd0a0 and url nats://localhost:4222
```
